blockchain/src/codec.rs

```diff
@@ -0,0 +1,388 @@
+            5 => Ok(MessageType::GetMempoolTxs),
+            _ => Err(ReadError::Custom(
+                format!("unknown message type: {}", value).into(),
+            )),
+        }
+    }
+}
+
+impl Encodable for BlockHeader {
+    fn encode(&self, w: &mut impl Writer) -> Result<(), WriteError> {
+        w.write_u64(b"version", self.version)?;
+        w.write_u64(b"height", self.height)?;
+        w.write_blockid(b"prev", &self.prev)?;
+        w.write_u64(b"timestamp_ms", self.timestamp_ms)?;
+        w.write_hash(b"txroot", &self.txroot)?;
+        w.write_hash(b"utxoroot", &self.utxoroot)?;
+        w.write_u8_vec(b"ext", &self.ext)?;
+        Ok(())
+    }
+
+    fn encoded_length(&self) -> usize {
+        8 + 8 + 32 + 8 + 32 + 32 + 4 + self.ext.len()
+    }
+}
+
+impl Decodable for BlockHeader {
+    fn decode(buf: &mut impl Reader) -> Result<Self, ReadError> {
+        Ok(BlockHeader {
+            version: buf.read_u64()?,
+            height: buf.read_u64()?,
+            prev: buf.read_blockid()?,
+            timestamp_ms: buf.read_u64()?,
+            txroot: buf.read_hash()?,
+            utxoroot: buf.read_hash()?,
+            ext: buf.read_u8_vec()?,
+        })
+    }
+}
+
+impl Encodable for Inventory {
+    fn encode(&self, w: &mut impl Writer) -> Result<(), WriteError> {
+        w.write_u64(b"version", self.version)?;
+        self.tip.encode(w)?;
+        w.write_signature(&self.tip_signature)?;
+        w.write_u64(b"shortid_nonce", self.shortid_nonce)?;
+        w.write_shortid_vec(b"shortid_list", &self.shortid_list)?;
+        Ok(())
+    }
+
+    fn encoded_length(&self) -> usize {
+        8 + self.tip.encoded_length() + 32 + 8 + 4 + self.shortid_list.len()
+    }
+}
+
+impl Decodable for Inventory {
+    fn decode(buf: &mut impl Reader) -> Result<Self, ReadError> {
+        Ok(Inventory {
+            version: buf.read_u64()?,
+            tip: BlockHeader::decode(buf)?,
+            tip_signature: buf.read_signature()?,
+            shortid_nonce: buf.read_u64()?,
+            shortid_list: buf.read_shortid_vec()?,
+        })
+    }
+}
+
+// Tx now implement Encoder, but it's implementation isn't suitable for
+// encoding in that case
+fn read_tx(buf: &mut impl Reader) -> Result<Tx, ReadError> {
+    let tx_size = buf.read_u32()? as usize;
+    let vec = buf.read_vec(tx_size)?;
+    Tx::from_bytes(&vec).map_err(|_| ReadError::InvalidFormat)
+}
+
+fn write_tx(tx: &Tx, dst: &mut impl Writer) -> Result<(), WriteError> {
+    dst.write_u32(b"tx length", tx.encoded_length() as u32)?;
+    tx.encode(dst)
+}
+
+fn read_block_tx(src: &mut impl Reader) -> Result<BlockTx, ReadError> {
+    let tx = read_tx(src)?;
+    let len = src.read_u32()? as usize;
+    let proofs = src.read_vec_with(len, 1, |src| match src.read_u8()? {
+        0 => Ok(Proof::Transient),
+        1 => merkle::Path::decode(src)
+            .map(Proof::Committed)
+            .map_err(|_| ReadError::InvalidFormat),
```

Comment:
doesn't `Path::decode` return ReadError already?

```suggestion
,
```

Reply:
https://github.com/p0lunin/slingshot/blob/transfer_p2p_coding_to_readwriter/zkvm/src/merkle.rs#L367 no, it's return VMError

Reply:
oh, dang, gotta update that one two.